addons/mrp/models/mrp_production.py

```diff
@@ -516,6 +517,18 @@ def _compute_show_lot_ids(self):
         ('qty_positive', 'check (product_qty > 0)', 'The quantity to produce must be positive!'),
     ]
 
+    @api.depends('product_uom_qty', 'date_planned_start')
+    def _compute_forecasted_issue(self):
+        for order in self:
+            warehouse = order.location_dest_id.get_warehouse()
+            order.forecasted_issue = False
+            if order.product_id:
+                virtual_available = order.product_id.with_context(warehouse=warehouse.id, to_date=order.date_planned_start).read(['virtual_available'])[0]["virtual_available"]
```

Comment:
you should group product by warehouse and date to minimize the call to read (call read on all product that have the same date and the same warehouse)

Reply:
@amoyaux Do yo mean to do something like this? I don't see how to avoid the ugly triple loop product/warehouse/date without creating duplicated calls to read
```python
    @api.depends('product_uom_qty', 'date_planned_start')
    def _compute_forecasted_issue(self):
        self.forecasted_issue = False
        products = self.mapped('product_id')
        for product in products:
            orders_product = self.filtered(lambda order: order.product_id == product)
            dates = orders_product.mapped('date_planned_start')
            for date in dates:
                orders_date = orders_product.filtered(lambda order: order.date_planned_start == date)
                warehouses = orders_date.location_dest_id.get_warehouse(limit=None)
                for warehouse in warehouses:
                    orders_wh = orders_date.filtered(lambda order: order.location_dest_id.get_warehouse() == warehouse)
                    virtual_available = product.with_context(warehouse=warehouse.id, to_date=date).read(['virtual_available'])[0]["virtual_available"]
                    for order in orders_wh:
                        virtual_available_order = virtual_available
                        if order.state == 'draft':
                            virtual_available_order = virtual_available + order.product_uom_qty
                        if virtual_available_order < 0:
                            order.forecasted_issue = True
```

Reply:
Ok let keep it like that

Reply:
@amoyaux Like the previous code or with the triple loop?

Reply:
Like the previous code